src/Utils/kernel_launching.jl

```diff
@@ -6,10 +6,27 @@ using KernelAbstractions
 using Oceananigans.Architectures
 using Oceananigans.Grids
 
-const MAX_THREADS_PER_BLOCK = 256
-
 flatten_reduced_dimensions(worksize, dims) = Tuple(i ∈ dims ? 1 : worksize[i] for i = 1:3)
 
+const MAX_GPU_THREADS_PER_BLOCK = 256
+
+function heuristic_workgroup(arch, Nx, Ny, Nz)
+    # One-dimensional column models:
+    Nx == 1 && Ny == 1 return (1, 1)
+
+    # Two-dimensional y-z slice models:
+    Nx == 1 && return (1, min(MAX_GPU_THREADS_PER_BLOCK, Ny))
+
+    # Two-dimensional x-z slice models:
+    Ny == 1 && return (1, min(MAX_GPU_THREADS_PER_BLOCK, Nx))
+
+    # Three-dimensional GPU models
+    arch isa GPU && return (Int(√MAX_GPU_THREADS_PER_BLOCK), Int(√MAX_GPU_THREADS_PER_BLOCK))
+
+    # Three-dimensional CPU models
+    return (Nx, Ny)
```

Comment:
@glwagner I am trying to see why `return (Nx, Ny)` is the formula. 

Reading code here https://github.com/JuliaGPU/KernelAbstractions.jl/blob/545457ad3918b4d5aa687d6423f3e33909f368b6/src/KernelAbstractions.jl#LL348-L354 and below I got the impression that the workgroup size is the "granularity" that each thread will get. 

So if my granularity is Nx,Ny and I launch a kernel with worksize Nx,Ny,Nz I will spawn to at most Nz threads?  For some cases that could be 1 or some small number? 

To be fully general does it need to be Nx,Ny/NTHREADS or something more complicated involving Nz to ensure your workgroup size is < "data size"/NTHREADS.   Where "data size" is Nx*Ny*Nz - or it could be my reading of https://juliagpu.github.io/KernelAbstractions.jl/dev/ is confused!

For problems which mix 3d and 2d multithreading it might also be that some extra either dynamic or dispatch based variants are needed?

Reply:
> @glwagner I am trying to see why return (Nx, Ny) is the formula.

Only because it improved one benchmark, so it's worth discussing...

I think see what you're saying. Do you think we would want something like:

```julia
Nthreads = Base.Threads.threads()

y_residual_threads = max(1, Nthreads - Nz - 1)
By = floor(Int, Ny / y_residual_threads)

x_residual_threads = max(1, y_residual_threads - Ny - 1)
Bx = floor(Int, Nx / y_residual_threads)

return (Bx, By)
```

Thus for example if `Nz = 1`, then the problem is partitioned along `y`.

We could also add an explicit case for `Nz=1` and design an algorithm robust for `Nz > 1`.

@christophernhill your link may not have gone to the intended code block in `KernelAbstractions` --- can you send again?

@vchuravy maybe you have a recommendation?

Reply:
@glwagner I think we need @vchuravy to clarify what WorkGroupSize means. 

The link is to where it is used, but it is hard to follow exactly how it then impacts the scheduling of CPU threads/GPU thread-blocks - so I may be misreading the code. My conclusion was that on a CPU `WorkGroupSize` ends up being the the size of the "blocks" the range is subdivided into for passing into different threads. I haven't tested that theory, so I could be wrong.

Reply:
The link goes to this docstring:

```
"""
    Kernel{Device, WorkgroupSize, NDRange, Func}
Kernel closure struct that is used to represent the device
kernel on the host. `WorkgroupSize` is the number of workitems
in a workgroup.
"""
```

Do you mean that the tuple is added to the type signature of `Kernel`?

Reply:
> @glwagner I think we need @vchuravy to clarify what WorkGroupSize means.
> 
> The link is to where it is used, but it is hard to follow exactly how it then impacts the scheduling of CPU threads/GPU thread-blocks - so I may be misreading the code. My conclusion was that on a CPU `WorkGroupSize` ends up being the the size of the "blocks" the range is subdivided into for passing into different threads. I haven't tested that theory, so I could be wrong.

Does that explain why we get speed up for 512^3 if we set (512, 512)? Basically this translates into splitting up the outer dimension?

Reply:
Yeah - if my interpretation is correct then 512^3 would be fine, but 512x512x1 might be a problem. If we can clarify what `WorkGroupSize` means then coming up with new heuristic, if needed, should be straight forward.

Reply:
I see @vchuravy said "yes" on Slack - so I think that means a 512x512x1 (or 2,3 etc...) problem would not scale without tweaking the heuristic.

Reply:
Here's my 🕵️ 

`workgroupsize` is used on the CPU here:

https://github.com/JuliaGPU/KernelAbstractions.jl/blob/51a59dc8ff2427b53db93a2e50efb4a972e370f5/src/cpu.jl#L106

to calculate `iterspace`. `iterspace` is used in `__run`:

https://github.com/JuliaGPU/KernelAbstractions.jl/blob/51a59dc8ff2427b53db93a2e50efb4a972e370f5/src/cpu.jl#L116-L137

possibly the logic in `__run` would limit us to one thread if `Nz=1`?

```
 N = length(iterspace)
    Nthreads = Threads.nthreads()
    if Nthreads == 1
        len, rem = N, 0
    else
        len, rem = divrem(N, Nthreads)
    end
    # not enough iterations for all the threads?
    if len == 0
        Nthreads = N
        len, rem = 1, 0
    end
```

Reply:
Each task iterates over a number of blocks. The number of blocks is given by splitting the ndrange by the provided workgroupsize.